web-js/src/components/RealmApp.js

```diff
@@ -0,0 +1,39 @@
+import React from "react";
+import * as Realm from "realm-web";
+
+const RealmAppContext = React.createContext(null);
+
+export function RealmAppProvider({ appId, children }) {
```

Comment:
Does 'RealmAppProvider' need to exist for this application? It seems it's functionality is basically: 

* logIn
* logOut
* pass logged in user to other components
* pass the Realm App to other components. 

An alternative approach to a 'RealmAppProvider' component is to do the following:
* The actual logIn/logOut could be methods on a component (probably WelcomePage and App for this application)
* You can get the logged-in user by calling '[Realm.App.currentUser]'(https://docs.mongodb.com/realm-sdks/js/latest/Realm.App.html#currentUser) anywhere in your application. 
* You can get the Realm App anywhere in your application by calling '[Realm.App.getApp](https://docs.mongodb.com/realm-sdks/js/latest/Realm.App.html#getApp)' 


This alternative approach offers simplicity (which is a major goal of this project since it is a template app and users may be confused by any additional/unnecessary bulk), and parity with the react native template app).

You can see the implementation for React Native here: https://github.com/mongodb-university/realm-template-apps/blob/main/react-native/src (let me know if you have questions)

Reply:
Does it strictly _need_ to exist? No. But it does include a few advantages when you view the template as a base app devs can build from:

- **Singleton definition & instantiation** - the RN template instantiates the App several times (once per view)
- **Composability & reuse** - some folks use multiple Realm apps and this approach lets you dynamically define multiple instead of hardcoding one app at the module level.
- **State encapsulation** - the RN template mirrors the `user` as state in `WelcomeView` and would need to do this for any component where auth changes should affect app state (i.e. really any auth change). Along the lines of the singleton point, this encapsulates the state update in a single spot.
- **Future-poofing** - our tentative plan is to use the provider/consumer approach in the official Realm hooks, so this API should hopefully not require significant changes to adapt the official package whenever we release it.

Reply:
Honestly, I can see the validity in both arguments.
The root cause of this, is most likely that the Realm Web API is too simple to be really useful without introducing bugs from components not rendering when a user logs in / out.

I see this component as a workaround for an API which is missing important features. This makes it harder for users to consume the app template, but at least we're not teaching users a fast route to trouble.

Reply:
Just putting my thoughts here.  I agree the provider pattern is the way to go here, but I also think we should be provided from `realm-js` so that the developer doesn't have to write it themselves.